imports/composite_redolog_manager.go

```diff
@@ -0,0 +1,330 @@
+		return s.storeFunc(batch, redoLogFile, batchOffset, skipBackfillRows)
+	}
+	return nil
+}
+
+// applyUpsertBatchWithLogging is used for normal ingestion (kafka or http)
+func (s *CompositeRedologManager) applyUpsertBatchWithLogging(batch *common.UpsertBatch, offsetInSource int64) error {
+
+	utils.GetReporter(s.Table, s.Shard).GetCounter(utils.IngestedUpsertBatches).Inc(1)
+	utils.GetReporter(s.Table, s.Shard).GetGauge(utils.UpsertBatchSize).Update(float64(len(batch.GetBuffer())))
+
+	s.Lock()
+	defer s.Unlock()
+
+	s.msgReceived++
+	redoLogFile, batchOffset := s.AppendToRedoLog(batch, offsetInSource)
+	if s.storeFunc != nil {
+		return s.storeFunc(batch, redoLogFile, batchOffset, false)
+	}
+	return nil
+}
+
+// WaitForRecoveryDone block call to wait for recovery finish
+func (s *CompositeRedologManager) WaitForRecoveryDone() {
+	select {
+	case <-s.readyChan:
+		break
+	}
+
+	// report redolog size after replay
+	utils.GetReporter(s.Table, s.Shard).GetGauge(utils.NumberOfRedologs).Update(float64(s.GetNumFiles()))
+	utils.GetReporter(s.Table, s.Shard).GetGauge(utils.SizeOfRedologs).Update(float64(s.GetTotalSize()))
+}
+
+// WriteUpsertBatch append upsertbatch to redolog and save to storage
+func (s *CompositeRedologManager) WriteUpsertBatch(batch *common.UpsertBatch) error {
+	if s.fileRedoLogManager != nil {
+		return s.applyUpsertBatchWithLogging(batch, -1)
+	} else {
+		return errors.New("Local file redolog not configured")
+	}
+}
+
+// AppendToRedoLog append upsert batch into redolog file or commit offset
+func (s *CompositeRedologManager) AppendToRedoLog(upsertBatch *common.UpsertBatch, offsetInSource int64) (int64, uint32) {
+	if s.fileRedoLogManager != nil {
+		if s.kafkaReader != nil && offsetInSource >= 0 {
+			// when save redolog in file and ingestion is from kafka, need to remember kafka offset separately
+			s.kafkaReader.Commit(offsetInSource)
+		}
+		return s.fileRedoLogManager.AppendToRedoLog(upsertBatch, offsetInSource)
+	} else {
+		return s.kafkaReader.AppendToRedoLog(upsertBatch, offsetInSource)
+	}
+}
+
+// UpdateMaxEventTime update max event time for related redolog file
+func (s *CompositeRedologManager) UpdateMaxEventTime(eventTime uint32, redoFile int64) {
+	if s.fileRedoLogManager != nil {
+		s.fileRedoLogManager.UpdateMaxEventTime(eventTime, redoFile)
+	} else {
+		s.kafkaReader.UpdateMaxEventTime(eventTime, redoFile)
+	}
+}
+
+// CheckpointRedolog clean up obsolete redolog files and save checkpoint offset
+func (s *CompositeRedologManager) CheckpointRedolog(cutoff uint32, redoFileCheckpointed int64, batchOffset uint32) error {
+	if s.fileRedoLogManager != nil {
+		return s.fileRedoLogManager.CheckpointRedolog(cutoff, redoFileCheckpointed, batchOffset)
+	} else {
+		return s.kafkaReader.CheckpointRedolog(cutoff, redoFileCheckpointed, batchOffset)
+	}
+}
+
+func (s *CompositeRedologManager) GetTotalSize() int {
+	if s.fileRedoLogManager != nil {
+		return s.fileRedoLogManager.GetTotalSize()
+	} else {
```

Comment:
so many duplicated if else logic, I think you can encapsulate an interface on top of s.fileRedoLogManager and kafkaReader, and setting the kafkaReader only at boostrap time